File: dcp-solutions/2025-01-30/solution.c
```c
#include <stdio.h>
#include <stack>

int is_balanced(const char *s) {
    stack<char> st;
    for (const char *p = s; *p != '\0'; ++p) {
        if (*p == '(' || *p == '*') {
            st.push(*p);
        } else if (*p == ')') {
            if (st.empty()) return 0;
            if (st.top() == '*' || st.top() == '(') {
                st.pop();
            } else if (st.top() == '*') {
                while (!st.empty() && st.top() != '(') {
                    st.pop();
                }
                if (st.empty() || st.top() != '(') return 0;
                st.pop();
            }
        }
    }
    return st.empty();
}
```

File: dcp-solutions/2025-01-30/solution.rb
```ruby
def check_balance(s)
  stack = []
  mapping = { '?' => '' }

  def get_mapping(c)
    mapping[c] || c
  end

  s.each_char do |c|
    if c == '('
      stack << c
    elsif c == ')'
      return false unless stack.empty? || get_mapping(stack.pop) == '('
    elsif c == '*'
      stack << get_mapping(c)
    end
  end

  stack.empty?
end
```